ibis/impala/udf.py

```diff
@@ -336,6 +341,31 @@ def _ibis_string_to_impala(tval):
         return repr(result)
 
 
+_COMPLEX_RE = re.compile(r'(struct|map|array)\<([\s\S]+)\>$')
+_STRUCT_FIELD_RE = re.compile(
+    r"`?(\w+)`?:((?:struct|map|array)\<[\s\S]+?\>|\S+)(?:\s+comment '([^']+)')?(?:,\n|$)") # noqa E501
+_MAP_FIELD_RE = re.compile(r"([^,]+),((?:struct|map|array)\<[\s\S]+?\>|.*)")
+
+
+def _impala_complex_type_to_ibis(tval):
```

Comment:
This function needs some unittests that test both valid and invalid input.

Reply:
I was a touch confused by your test case layout and the fact i was unable to get the udfs tests working.  Where is the best place to implement these tests, do i create the additional types in ImpalaEnv or do i create test cases particularly for this?

Reply:
How are you making sure that this code works if you're unable to test?

Reply:
I am using the work i am doing here for another project which has access to full Cloudera cluster so i know it works, plus i tested it manually.  I just needed someone to point me in the right direction on where to put the tests

Reply:
It doesn't look like there are actually any tests that directly test the type conversion without udfs.  I have added a test case for the udfs by adding another table to the MockConnection and in the TestWrapping class.  I am able to test this class as it doesn't use the dockerized impala but anything else crashes it.